File: src/TaskOTime/TaskOTime.TimeTrackingServices/TimeItemByEventTimeComparer.vb
Imports System.Collections.Generic

Namespace ActiveDevelop.TimeTrackingServices
    Public Class TimeItemByEventTimeComparer(Of IndexType As {Structure, IComparable(Of IndexType)}, TimeItemType As {Class, ITimeItem(Of IndexType), New})
        Implements IComparer(Of TimeItemType)

        Public Property IgnoreOnce As TimeItemType

        Public Function Compare(x As TimeItemType, y As TimeItemType) As Integer Implements IComparer(Of TimeItemType).Compare
            If IgnoreOnce IsNot Nothing AndAlso x IsNot Nothing AndAlso Object.Equals(x.IDTimeItem, IgnoreOnce.IDTimeItem) Then
                IgnoreOnce = Nothing
                Return -1
            End If

            Dim xEventTime As DateTimeOffset? = If(x Is Nothing, CType(Nothing, DateTimeOffset?), x.EventTime)
            Dim yEventTime As DateTimeOffset? = If(y Is Nothing, CType(Nothing, DateTimeOffset?), y.EventTime)

            If Not xEventTime.HasValue AndAlso Not yEventTime.HasValue Then
                Return 0
            End If

            If Not xEventTime.HasValue Then
                Return -1
            End If

            If Not yEventTime.HasValue Then
                Return 1
            End If

            Return xEventTime.Value.CompareTo(yEventTime.Value)
        End Function
    End Class
End Namespace

File: src/TaskOTime/TaskOTime.TimeTrackingServices/TimeItemByEventTimeComparer.cs
using System;
using System.Collections.Generic;

namespace ActiveDevelop.TimeTrackingServices
{
    public class TimeItemByEventTimeComparer<IndexType, TimeItemType> : IComparer<TimeItemType>
where IndexType : struct, IComparable<IndexType>
where TimeItemType : class, ITimeItem<IndexType>, new()
    {

        public TimeItemType IgnoreOnce { get; set; }

        public int Compare(TimeItemType x, TimeItemType y)
        {
            if (IgnoreOnce is not null && x is not null && Equals(x.IDTimeItem, IgnoreOnce.IDTimeItem))
            {
                IgnoreOnce = null;
                return -1;
            }

            var xEventTime = x is null ? default : x.EventTime;
            var yEventTime = y is null ? default : y.EventTime;

            if (!xEventTime.HasValue && !yEventTime.HasValue)
            {
                return 0;
            }

            if (!xEventTime.HasValue)
            {
                return -1;
            }

            if (!yEventTime.HasValue)
            {
                return 1;
            }

            return xEventTime.Value.CompareTo(yEventTime.Value);
        }
    }
}